arangod/RocksDBEngine/RocksDBMetaCollection.cpp

```diff
@@ -450,114 +458,304 @@ std::unique_ptr<containers::RevisionTree> RocksDBMetaCollection::computeRevision
   return buildTreeFromIterator(it);
 }
 
-bool RocksDBMetaCollection::needToPersistRevisionTree(rocksdb::SequenceNumber maxCommitSeq) const {
-  if (!_logicalCollection.useSyncByRevision()) {
-    return maxCommitSeq > _revisionTreeApplied.load();
+Result RocksDBMetaCollection::takeCareOfRevisionTreePersistence(
+    LogicalCollection& coll, RocksDBEngine& engine,
+    rocksdb::WriteBatch& batch, rocksdb::ColumnFamilyHandle* const cf,
+    rocksdb::SequenceNumber maxCommitSeq,
+    bool force, std::string const& context, std::string& scratch,
+    rocksdb::SequenceNumber& appliedSeq) {
+
+  TRI_ASSERT(coll.useSyncByRevision());
+
+  // might lower `appliedSeq`!
+
+  // Get lock on revision tree:
+  std::unique_lock<std::mutex> guard(_revisionTreeLock);
+
+  maxCommitSeq = _meta.committableSeq(maxCommitSeq);
+  if (needToPersistRevisionTree(maxCommitSeq, guard)) {
+    rocksdb::SequenceNumber seq = maxCommitSeq;
+
+    if (seq < _revisionTreeApplied) {
+      // this is a special case when we have an applied sequence number
+      // which is higher than what we have blockers for. this can only
+      // happen during tree rebuilding. in this case we have to bump the
+      // seq number forward, because we cannot generate a tree from the past.
+      seq = _revisionTreeApplied;
+    }
+    scratch.clear();
+
+    try {
+      seq = serializeRevisionTree(scratch, seq, force, guard);
+    } catch (std::exception const& ex) {
+      LOG_TOPIC("33691", WARN, Logger::ENGINES)
+          << context << ": caught exception during revision tree serialization: " << ex.what();
+    
+      auto& engine =
+          _logicalCollection.vocbase().server().getFeature<EngineSelectorFeature>().engine<RocksDBEngine>();
+      auto* db = engine.db();
+      
+      rocksdb::WriteOptions wo;
+      
+      RocksDBKey key;
+      // remove revision tree from rocksdb
+      key.constructRevisionTreeValue(_objectId);
+      rocksdb::Status s = batch.Delete(cf, key.string());
+      
+      if (s.ok() || s.IsNotFound()) {
+        s = db->Write(wo, &batch);
+      } 
+      if (!s.ok()) {
+        LOG_TOPIC("af3de", WARN, Logger::ENGINES)
```

Comment:
As far as I understand this will eventually heal itself. The only problem is if we crash in the meantime, then the new instance after the restart could continue with a wrong tree and this could potentially only be detected randomly. But what can we do? FATAL exit would be enforcing a crash, which is the exact worse case. So I think this is correct behaviour.

Reply:
Can't we just opt out of this specific method and retry again later including the removal of the document?
afaik the Tree we have is okay or not?